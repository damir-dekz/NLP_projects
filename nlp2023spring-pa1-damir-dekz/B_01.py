import os
import shutil
import zipfile
import pandas as pd
import re

# DO NOT CHANGE
data_path = "./data/bcc"
zip_path = './data/bbc-fulltext.zip'


# DO NOT CHANGE
def prepare_files():
    with zipfile.ZipFile(zip_path, 'r') as zip_ref:
        zip_ref.extractall(data_path)
    # Define the file extension to look for
    file_extension = ".txt"
    # Initialize an empty list to store the file paths
    file_paths = []
    # Use os.walk to traverse through the directory and subdirectories
    for dirpath, dirnames, filenames in os.walk(data_path):
        for filename in filenames:
            if filename.endswith(file_extension):
                # Construct the full file path
                file_path = os.path.join(dirpath, filename)
                # Append the file path to the list
                file_paths.append(file_path)
    return file_paths


def find_proper_names(text):
    result_name = re \
        .findall(
        '(?:like|firms)?\s(?:[A-Z]\w+|the)(?:\s[A-Z]\w+(?:\'s))+'
        '|(?:like|firms)?\s(?:[A-Z]\w+)(?:\s[A-Z]\w+)+'
        '|(?:[A-Z]+){2,4}'  # ex: AA,AAA,AAAA
        '|Mr [A-Z]\w+'
        '|(?:[A-Z]\w+ )*(?:[A-Z][a-z]+)*\s(?:[A-Z]\')*[A-Z][a-z]+\'s' # ex: Alan Bennett's, Amazon's, Van Doren Stern's
        '|[A-Z]\'[A-Z]\w+'
        '|(?:legend|leader)\s\w+(?:\s\w+)?'
        '|\s[0-9][0-9](?:\s[A-Z][a-z]+){2}'
        '|(?:[A-Z][a-z]+)+\sand\s(?:[A-Z][a-z]+)+'
        '|(?:[A-Z]|\$[0-9]+)\.(?:[A-Z]\.|[0-9][a-z]+)'
        '|\s(?:in|on)\s[A-Z]\w+'
        '|[A-z]\w+\.com'
        '|(?:[Ff]or|at)\s[A-Z]\w+'
        '|the(?:\s[A-Z]\w+){2,}'
        , text)

    # result_name2 = re.findall('(?:[A-Z][a-z]+)*\s(?:[A-Z]\')*[A-Z][a-z]+\'s', text)
    result_name = [name
                   .replace("legend ", "")
                   .replace("and ", "")
                   .replace("And ", "")
                   .replace(" in ", "")
                   .replace(" on ", "")
                   .replace("$", "")
                   .replace("leader ", "")
                   .replace("like ", "")
                   .replace("for ", "")
                   .replace("at ", "")
                   # .replace('""', "")
                   for name in result_name]
    result_name = [re
                   .sub(r'((?:[A-Z][a-z]+)+)\sand\s((?:[A-Z][a-z]+)+)', r'\1\r\2', name)
                   for name in
                   result_name]
    result_name = [re
                   .sub(r'((?:[A-Z]\w+an))\s((?:[A-Z]\w+)(?:\s[A-Z]\w+)+)', r'\1\r\2', name)
                   for name in
                   result_name]
    # result_name2 = [re
    #                 .sub(r'([A-Z]\w+)\'s', r'\1', name)
    #                 for name in
    #                 result_name2]
    # result_name3 = result_name + result_name2
    return result_name


# Function to process a single file
def process_files(files):
    res = []
    unique_names = []
    for file_path in files:
        with open(file_path, "r", encoding='unicode_escape') as text_file:
            # Read the text from the file
            text = text_file.read()
            # TODO process test and extract names, write code below
            proper_names = find_proper_names(text)
            if proper_names:
                res.append(proper_names)
                for lists in res:
                    for items in lists:
                        unique_names.append((items.strip()))
    unique_names = list(set(unique_names))
    for i in range(len(unique_names)):
        z = re.match('(?:the )?(?:([A-Z]\w+)\'s)', unique_names[i])
        # z_the = re.match('(?:[T]he )?((?:[A-Z]\w+))', unique_names[i])
        if z:
            unique_names[i] = z.groups()[0]
        # elif z_the:
        #     unique_names[i] = z_the.groups()[0]
    # end code
    return unique_names


# DO NOT CHANGE
def safe_names_to_csv(unique_names):
    df = pd.DataFrame()
    df["names"] = unique_names
    df = df.drop_duplicates()
    df = df.sort_values(by="names")
    df.to_csv("./result/B_01.csv", index=False, header=False)


# DO NOT CHANGE
def main():
    files = prepare_files()
    unique_names = process_files(files)
    safe_names_to_csv(unique_names)
    shutil.rmtree(data_path, ignore_errors=False, onerror=None)


if __name__ == "__main__":
    main()
