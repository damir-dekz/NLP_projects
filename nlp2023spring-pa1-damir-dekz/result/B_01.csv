10 Downing Street
10.7bn
10.9bn
11.1bn
27.5bn
9.3bn
9.7bn
A.D.
AAA
AI
AOL
Abstract Rendering Architecture
Afghanistan
Africa
Agriculture Organisation
Airbus A321
Alan Beith
Alan Bennett
Alan Bennett's
Alan Greenspan
Alastair Campbell
Alexander Catwoman
Alexander Kosenkow
Alice Sebold's
Alistair Cragg
Allied Domecq
Allied Domecq's
Amanda Ross
Amazon
Amazon.com
AmericanAlan Webb
AmericanJustin Gatlin
Andersen
Anderson
Andrew McFadyen
Andrew Taylor's
Anniversary Special Awards
Apple
Apple Powerbook
April
Ashia Hansen
Asia
Askar Akaev
Aston Moore
Athens
Athens Olympic
Athletics Ireland
August
Aulis
Australia
AustralianCraig Mottram
BA
BBC
BBC News
BBC Radio
BMO
BNP Paribas
BT
Ballantine
Bank
Barbara Follet
Barbara Windsor
Beijing
Being Digital
Ben Wishaw
Berhane Adera
Bertelsmann
Birkbeck College
Birmingham
Birmingham GrPrix
Blackpool
Blackpool Hotel Association
Blair
Boston
Bournemouth Brighton
Brighton
Britain
British Airways
British Olympic
Brown
Brussels
Bulgaria Hungary
But David Frost
But Edward Leigh
But Greene
But Mr Balls
CBI
CD
CDP
Cabinet Office
Cambodia
Campbell
Capra
Carlos Saldarriaga
Carolina Kluft
Cathy Freeman
Cepes
Chancellor Gordon Brown
Charles Kennedy
Charles Wintour Award
China
Chinese
Chivas Regal Scotch
Choice Awards
Christian Andersen's
Christian Islamic
Christian Slater
Christine Arron
Christopher Oram
Circle Theatre Awards
Circle Whatsonstage
Clare
ClearView Economics
Cloud Atlas
Club World
Colin Asplin
Commonwealth European
Conference Centre
Connected World
Conservative Party
Copenhagen
Cory Doctorow
Coudenberg Cellars
Coudenberg Palace
Crown Prince Frederik
Crown Princess Mary
Cruel
Cuckoo
Cuyo
DOS
Dame Judi Dench
Dan Brown's
Dan Gillmor
Dark Materials
David Eldridge
David Mikosz
David Mitchell's
December
December January
Denmark
Deputy Prime Minister John Prescott
Diageo
Diana Rigg
Diane Allahgreen
Disney
Dominic Cooper
Donuts Baskin
Doren Stern
Dorset Torbay
Douglas Hodge
Dr Fox
Dresdner Kleinwort Wasserstein
Dublin
Dutchman Patrick
EOC
Early Warning System
East Anglia's
Ed Balls
Eddie Redmayne
Education Agriculture
Edward Albee
Electronic Fronter Foundation
Encounter
England
English Partnerships
Ethiopia
EuropeanIndoor Championships
European Commissioner
Eve
Eve Ensler
FAO
FTSE
Falling
False
Farmers
Federal Reserve
Festen
Financial Group
Finland
Finnigan
Fionnualla Britton
Flew Over
Food Agriculture
Foreign Secretary
Foreign Secretary Jack Straw
Former Liberal DemocrMP Jackie Ballard
France
Francis Obikwelu
Frank Capra's
Frank Stone
Frankie Fredericks
Friday
Funny Thing Happened
GMT
GMTV
Gardener Mark
Gateshead
Geraldine McCaughrean
Germany
Ghent
Gillian Shepherd
Glasgow
Global Information
Go Digital
Gold Dust
Google
Gordon
Gordon Brown
Graham Norton
Grassroots Media
Greece
Greene
Greenspan
Greenwich
Guy
Hans Christian Andersen
Harold Bloom
Harriet Harman
Harvey Keitel
Havana Club
Heizo Takenaka
Helena Christensen
Helsinki
Henri Josserand
Henry Moore
Herbert Hoover
Hildegard Bechtler
History Boys
Hugh Grant
IAAF
IE
IFES
IMF
ISBN
IT
Ian MacNeil
Ian Mackie
Ibrox
In August
In London
In October
In Serbia
Inc
India
India China
Indonesia Turkey
Information Act
Internet Explorer
Iphigenia
Ireland
Irrigation Users
It
Jacob
Jadel Gregorio's
James Stewart
Jamie Firestone
January
Japan
Jason Gardener
Jean Kalman
Jeremy Paxman
Jerry Springer
Jesus
Joanne Harris
John Guy
Joice Maduaka
Jolanda Ceplak
Jolene Byrne
Jon Thoday
Joseph O'Connor's
Julie Andrews
June
Karl Elsener's
Kelly Reilly
Ken Mayland
Kenenisa Bekele's
Kim Collins
Kim Gevaert
LVMH
Labor Department
Labour
Labour MP
Lamine Diack
Las Vegas
Last Summer
Laura Michelle Kelly
Lee Evans
Lehman
Leipzig
Leipzig International
Leonard Scott
Levy
Liam Fox
Lib Dem
Liberal DemocrSarah Teather
Liberal Democrat
Lievin
Little Lower
London
Lord Ashcroft
Luc Bondy
MD
MIT
MP
MP Kefa Ssempgani
MPC
MSN Messenger
Madeley Judy
Madonna
Madrid
Male MPs
Manchester
Marc Blume
Margaret Moran MP
Maria McCambridge
Mark Carroll
Mark Lewis
Martin Broughton
Martin McDonagh
Mary Magdalene
Mary Poppins
Mary Queen
Matthew Turnbull
Maurice
Maurice Greene
Meanwhile Lib Dem
Measure
Media
Media Labs
Melbourne
Menatep Group
Menatep Group's
Michael Howard
Michael Symmons Roberts
Microsoft
Microsoft Outlook
Mike Powell
Mikhail Khodorkovsky
Mikko Hypponen
Mobile PC
Monaco
Monday
Monetary Policy Committee
Mr
Mr Balls
Mr Beith
Mr Blair
Mr Blair's
Mr Brown
Mr Brown's
Mr Campbell
Mr Campbell's
Mr Chidambaram
Mr Doctorow
Mr Eddington
Mr Greenspan's
Mr Kennedy
Mr McFadyen
Mr Negroponte
Mr Saldarriaga
Mr Stone
Mr Straw
Mr Thoday
Mr Thomas
Ms Ensler
Ms Hewitt
My Heart
My Own
NAO
NIA
Nathan Lane
National Executive Committee
National Theatre
New
New Clothes
New Vision
New York
Nicholas Negroponte
Nick Van
Night Season
Norway
Norwich Union European Indoor
Norwich Union GrPrix
Not
Notes
Novelist Andrea Levy
November
Number One
O'Sullivan
OS
On Friday
On Monday Lib Dem
One
Or GPS
Owen McCafferty
PB
PC
PDA
Palaniappan Chidambaram
Palm Pilot
Pam Ferris
Pamela Travers
Paris
Parken
Parliament
Parliamentary Presidential
Patricia Hewitt
Paul Arditti
Paul Brown
Paul Rhys
Paul Sheard
Pele
Pernod
Pernod Ricard
Peru
Peter Mandelson
Poland
President
Prime Minister Tony Blair
Prince Frederik
Prof Joni Lovenduski
Prof Lovenduski
Professor Joni Lovenduski
Psion Organiser
QV
Qatar
Queen Mary
RSC
Rebecca Lenkiewicz
Richard
Richard Eyre
Richard Griffiths
Richard Judy
Richard Parsons
Richard Thomas
Richard Wentworth
Rick Egelton
Robbie Williams
Robert Sinche
Rod Eddington
Romola Garai
Rosneft
Rufus Norris
Russian
RussianIrina Shevchenko
SEC
Samuel Barnett
Sandra Gidley
Santry
Sarah Claxton
Sarah Mukasa
Saturday
Scotland
Secretary Patricia Hewitt
September
Sheffield
Simon McBurney
Sinclair QL
Sinclair Spectrum
Sir Cameron Mackintosh
Sir Digby
Sir Digby Jones
Sir John Bourn
Sir Richard Attenborough
Sir Roger Moore
Sir Trevor McDonald
Slovenia
Small Island
Smart Tags
Social
Sony
Sony Net Minidisc
Sony Walkman
South
South Africa
Soviet Republics
Spend Spend Spend
St
St Andrews
Stanley Townsend
Stephen Toulouse
Steve Brown
Stuttgart
Suddenly
Sunday
Susan Fletcher
Sweden
Sweeney Todd
Swiss Army Knife
Sybari Software
Sydney Olympic
Symmons Roberts
TBWA
TUC
TV
Tate Britain
That
The
The Agricultural Information Project
The American Boy
The Apple Powerbook
The AutoLink
The Bishop
The Board
The Casio
The Cepes
The Critics
The Da Vinci Code
The Darling Buds
The Department
The Emperor's
The Food
The Freedom
The G7
The Goat
The Greatest Gift
The History Boys
The IAAF
The Kansas
The Kyrgyz Republic
The Labor Department
The Labour Party
The Lib Dems
The Life
The Little Mermaid
The Lovely Bones
The Milton Shulman Award
The Motorola Startac
The NAO
The National Audit Office
The Olympic
The Opera
The Orange Expressionist
The Pillowman
The Producers
The Psion
The Reporters Without Borders
The Scottish Executive
The Shadow Secretary
The Sinclair Executive
The St Kitts
The TV
The Tivo
The Tories
The Tory
The US
The UgandanMedia Council
The Ugly Duckling
The Vagina Monologues
The WSJ
The West End
The White House
The Zafi
Theresa May
Thursday
Tim Osborne
Time Warner
Time Warner's
Tirunesh Dibaba
Tokyo
Tom Hanks
Tony Blair
Tory Treasurer Lord Ashcroft
Tracey Emin
Trade
Trade Industry
Trevor Beattie
Tudor England
Tuesday
UK
US
UV
Uganda
Ugandan NGOs
Ukraine Georgia
Vagina Monologues
Van Doren Stern
Victoria Hamilton
Vinci Code
Virgil Tacitus
Vulcan
Wales Northern
Warner
Wednesday
Well ThEnds Well
West End
Westminster
What
Whitbread
Whitbread Prize
Windows
Windows Messenger
Women MPs
Wonderful Life
World Championship
World Commonwealth
World Cross Country Championships
World Indoor Championships
Worlds
Xinhua News Agency
Yelena Isinbayeva
Yugansk
Yvette Cooper
the Agricultural Information System
the Athens Olympics
the BBC News
the BBC World Service
the Boston Indoor Games
the British Book Awards
the British Chambers
the British Library
the Bupa GreIrelRun
the Commonwealth Games
the Equal Opportunities Commission
the EuropeanIndoor Championships
the European Cup
the European Indoors
the Evening Standard Theatre Awards
the Federal Reserve
the Financial Times
the General Election
the German Embassy
the GreBritain
the Hans ChristianAndersen Award
the Hans ChristianAnderson School
the Huaral Valley
the Hutton Inquiry
the IAAF Council
the Kodak Brownie
the Kyrgyz President
the Labour Party
the Lib Dems
the Liberal Democrats
the Marine Chronometer
the Media Council
the Millennium Dome
the Millennium Exhibition
the Millennium Village
the National Theatre
the New Testament
the Nintendo GameBoy
the Northern Echo
the Norwich Union GrPrix
the Olympic Games
the Orange Prize
the River Senne
the Royal Court Theatre
the Rt Rev Dr Tom Wright
the Scottish Exhibition
the Scottish National Party
the Scottish Parliament
the Sinclair Spectrum
the Smart Tags
the Sony Walkman
the Soros Foundation
the South West
the Swiss Army Knife
the Toshiba Libretto
the US Federal Reserve
the US Securities Exchange Commission
the United Nations
the United States
the Wall Street Journal
the World Bank
the World Championships
the World Food Programme
