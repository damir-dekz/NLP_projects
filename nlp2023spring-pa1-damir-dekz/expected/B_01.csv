10 Downing Street
9.3bn
A.D.
AAA
AI
AOL
AOL Europe
Afghanistan
Africa
Airbus
Alan Beith
Alan Bennett
Alan Bennett's
Alan Greenspan
Alan Webb
Alastair Campbell
Alexander
Alexander Kosenkow
Alice Sebold's
Alistair Cragg
Allied
Allied Domecq
Amanda Ross
Amazon
Amazon.com
American
Andersen
Anderson
Anderson School
Andrea Levy
Andrew McFadyen
Andrew Taylor's
Apple
ArtWorks
Ashcroft
Ashia Hansen
Asia
Askar Akaev
Aston Moore
Athens
Athens Olympic
Athletics Ireland
Ato
Aulis
Australia
Australian
AutoLink
AutoLinks
BA
BBC
BBC News
BBC Radio
BBC Radio 4's
BBC2's Newsnight
BMO Financial Group
BT
Ballantine
Bank of America
Bankash-A Trojan
Barbara Follet
Barbara Windsor
Basingstoke
Beijing
Beith
Bekele
Belgium
Ben Wishaw
Bennett
Berhane Adera
Bertelsmann
Betty
"Betty"""
Biblical
Birkbeck College
Birmingham
Birmingham Grand Prix
Bishop
Blackberry
Blackpool
Blackpool Hotel Association
Blair
Boston
Bournemouth
Brazil
Brazilian
Breville
Brighton
Britain
British
British Airways
British Olympic
Britons
Brown
Brussels
Budget
Bulgaria
Bush
CBI
CDP-101
Cabinet
Cabinet Office
Calico
Cambodia
Cambridge
Cameron Mackintosh
Cameron Mackintosh's
Campbell
Carfax
Carlos Saldarriaga
Carroll
Casio
Cathy Freeman
Catwoman
Ceplak
Chamber
Channel 4
Charles Kennedy
Chidambaram
China
Chinese
Christian
Christian Andersen
Christian Andersen's
Christian Slater
Christianity
Christine Arron
Christmases
Christopher Oram
Church
Cicero
Clare College
Claxton
ClearView Economics
Cloud Atlas
Club World
Cobh
Colchester
Colin Asplin
Collins
Commons
Commonwealth
Connected World
Conservatives
Copenhagen
Corpus
Cory Doctorow
Coudenberg Cellars
Coudenberg Palace
Courvoisier
Cragg
Craig Mottram
Crown Prince Frederik
Cuyo
DOS
Da Vinci Code
Dan Brown
Dan Brown's
Dan Gillmor
Danish
David Eldridge
David Frost
David Mikosz
David Mitchell's
Democrat
Democrats
Dems
Denmark
Diageo
Diamound Multimedia
Diana Rigg
Diane Allahgreen
Dibaba
Digby
Digby Jones
Digital
Disney
Dominic Cooper (His Dark Materials
Dorset
Douglas Hodge
Dr Fox
Dresdner Kleinwort Wasserstein
Dublin
Dunkin' Donuts
Dutchman Patrick van Balkom
EOC
Ed Balls
Eddie Redmayne
Eddington
Education
Edward Albee
Edward Leigh
Electronic Fronter Foundation
England
English
Ethiopia
Ethiopian
Ethiopians
European
European Indoor Championships
Eve Ensler
Eve Green
F-Secure's
FAO
FTSE
Fagin
False Servant
Federal Reserve
Ferris
Festen
Finland
Fionnualla Britton
Flipstart
Focus on
Fox
France
Francis Obikwelu
Frank Capra's
Frank Stone
Frankie Fredericks
French
G20
GMT
GMTV
Gateshead
Gatlin
Georgia
Geraldine McCaughrean
German
Germany
Ghent
Gillian Shepherd
Glasgow
Glenmorangie
Global Information
Gold Dust
Google
Gordon Brown
Gordon Brown's
Government
Graham Norton
Grassroots Media
Great Britain
Greece
Greene
Greenspan
Greenwich
Guy
Guy's My Heart
Hans
Hans Christian Andersen
Hansen
Harold Bloom
Harriet Harman
Harvey Keitel
Havana Club
Heizo Takenaka
Helena Christensen
Helsinki
Henri Josserand
Henry Moore
Herbert Hoover
Hildegard Bechtler
Hitler
Huaral
Hugh Grant
Hungary
IAAF
IFES
IMF
Ian MacNeil
Ian Mackie
Ibrox
India
Indian
Indonesia
Iphigenia
Ireland
Irish
Islamic
Italy
Jack Straw
Jackie Ballard
Jacob
Jadel Gregorio's
Jamaican
James Stewart
Jamie Firestone
Japan
Jarre
Jason Gardener
Jean Kalman
Jean-Michel Jarre
Jeremy Paxman
Jerry Springer - The
Jesus
Joanne Harris
John Bourn
John Guy
John Prescott
Joice Maduaka
Jolanda Ceplak
Jolene Byrne
Jon Thoday
Joni Lovenduski
Joseph O'Connor's
Judy
Judy Finnigan's
Julie Andrews
Justin Gatlin
Kampala
Kansas
Karl Elsener's
Kefa Ssempgani
Kelly Reilly
Ken Mayland
Kenenisa Bekele's
Kennedy
Kim Collins
Kim Gevaert
Kluft
Kodak
Kyrgyz
Labor Department
Labour
Labour MP
Lamine Diack
Langlands & Bell
Las Vegas
Laura Michelle Kelly
Lee Evans
Lehman Brothers
Leicester
Leipzig
Leonard Scott
Levy
Lewis-Francis
Liam Fox
Lib Dem
Liberal Democrats
Lievin
Lima
Linux
Liverpool
London
Luc Bondy
MIT
MPC
MSN Messenger
Madeley
Madonna
Madrid
Maine
Malibu
Manchester
Manhattan
Marc Blume
Margaret Moran
Maria McCambridge
Mark Carroll
Mark Lewis-Francis
Martin Broughton
Martin McDonagh
Mary Magdalene
Mary Poppins
Mary Poppins'
Mary Queen
Matthew Turnbull
Maurice
Maurice Greene
McFadyen
Media Labs
Media Player
Melbourne
Menatep Group
Menatep Group's
Michael Howard
Michael Symmons Roberts'
Microsoft
Microsoft Outlook
Mike Powell
Mikhail Khodorkovsky
Mikko Hypponen
Millennium Dome
Mobile
Monaco
Monetary Policy Committee
Moore
Moscow
Motorola Startac
Mr Balls
Mr Blair
Mr Blair's
Mr Brown
Mr Stone
Ms Ensler
Ms Hewitt
NAO
NIA
Namibian
Nathan Lane
National Executive Committee
National Theatre
Nationals in Santry
Negroponte
NetMD
New Clothes
New Vision
New York
Newton
Nicholas Negroponte
Nick Van den Brul
Nintendo
Nokia
Northern Ireland
Norway
Norwich Union European Indoor
Norwich Union Grand Prix
O'Sullivan
Odense
Olympic
Opera
Orange Expressionist
Owen McCafferty
PDA
Palaniappan Chidambaram
Pam Ferris
Pamela Travers
Paramount
Paris
Parliament
Parliamentary
Patricia Hewitt
Paul Arditti
Paul Brown
Paul Rhys
Paul Sheard
Pele
Pernod
Pernod Ricard
Peru
Peter Mandelson
Philip van Doren Stern
Pillowman
Poland
Poppins
Portugal
Producers
Prof Joni Lovenduski
Prof Lovenduski
Pulsar
Qatar
Queen Mary
RSC
Rebecca Lenkiewicz
Reuters
Richard
Richard Attenborough
Richard Eyre
Richard Griffiths
Richard Madeley
Richard Parsons
Richard Thomas
Richard Wentworth
Rick Egelton
Rings
Robbie Williams'
Robert Sinche
Rod Eddington
Roger Moore
Romola Garai
Ronco
Rosneft
Rufus Norris
Rufus Norris for Festen
Russia
Russian
Russian Irina Shevchenko
SEC
Saldarriaga
Sandra Gidley
Sarah Claxton
Sarah Mukasa
Sarah Teather
Scotland
Scott
Scottish
Sea
Seagram
Serbia
Sheffield
Shining City
Shrewsbury
Simon McBurney
Sinclair QL
Slovenia
Small Island
Smart Tags
Somerset
Sonia O'Sullivan
Sony
Sophos
South Africa
South America
Soviet
Soviet Republics
Spain
St Etienne
Stanley Townsend
State
Stephen Toulouse
Steve Brown
Stolichnaya
Stone
Straw
Stuttgart
Susan Fletcher
Sweden
Sweeney Todd and A Funny Thing Happened
Swiss Army Knife
Sybari Software
Sydney Olympic
Sylvia
Symmons Roberts
TBWA
TUC
Tacitus
Tamagotchi
Tandy
Tara
Tate
Tate Britain
The Agricultural Information Project for Farmers of the Chancay-Huaral Valley
The Board of Irrigation Users
The Department for Trade and Industry
The Food and Agriculture Organisation
The Kyrgyz Republic
The Labor Department
The Labour Party
The Milton Shulman Award
The Motorola Startac
The National Audit Office
The Reporters Without Borders
The Scottish Executive
The Sinclair Executive
The St Kitts
The Ugandan Media Council
The White House
Thomas
Tim Osborne
Time Warner
Time Warner's
TimeWarner
Times
Tirunesh Dibaba
Tivo
Tokyo
Tom Hanks
Tom Wright
Tony Blair
Torbay
Tories
Tory
Tory Treasurer
Tracey Emin
Treasury
Trevor Beattie
Trevor McDonald
Tudor
Turkey
UK
US
Uganda
Ugandan
Ukraine
Vagina Monologues
Van Doren Stern's
Victoria Hamilton
Virgil
Vulcan Inc's
WSJ
Wales
Warner Bros
West End
Westminster
Whatsonstage.com Theatregoers'
Whitbread
Whitbread Prize
Windows
Windows Messenger
World Cross Country Championships
Worlds
Xinhua News Agency
Yelena Isinbayeva
Yugansk
Yukos
Yvette Cooper
Zafi
Zafi-D
Zenith
a Conservative Party
first Dome
the Agricultural Information System
the Apple iPod
the Athletes' Commission
the BBC World Service
the Bank of England's
the Boston Indoor Games
the British Book Awards
the British Chambers of Commerce
the Equal Opportunities Commission
the European Indoor Championships
the European Indoors
the Federal Reserve
the Financial Times
the Food and Agriculture Organisation
the General Election
the German Embassy
the Hans Christian Andersen Award
the House of Commons
the Huaral Valley
the Hutton Inquiry
the IAAF Council
the Labour Party
the Labour Party's
the Liberal Democrats
the Marine Chronometer
the Media Council
the National Theatre
the New Testament
the Northern Echo
the Norwich Union Grand Prix
the Orange Prize
the Royal Court Theatre
the Scottish National Party
the Scottish Parliament
the Sinclair Spectrum
the Smart Tags
the Soros Foundation
the South West
the Swiss Army Knife
the Toshiba Libretto
the Trade and Industry
the US Federal Reserve's
the US Securities Exchange Commission
the United Nations
the United States
the University of East Anglia's
the University of St Andrews
the Wall Street Journal
the World Bank
the World Food Programme
tinny
"Â£30,000"
Â£9
